src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -76,26 +70,45 @@ trait Contexts { self: Analyzer =>
 
   var lastAccessCheckDetails: String = ""
 
-  /** List of symbols to import from in a root context.  Typically that
-   *  is `java.lang`, `scala`, and [[scala.Predef]], in that order.  Exceptions:
+  lazy val rootImportPackages: List[Symbol] = {
+    val paths =
+      if (settings.Ypredef.isSetByUser) {
+        if (settings.nopredef) error("Deprecated flag -Yno-predef isn't supported with -Ypredef")
+        if (settings.noimports) error("Deprecated flag -Yno-imports isn't supported with -Ypredef")
+        settings.Ypredef.value
+      }
+      else if (settings.noimports) Nil
+      else if (settings.nopredef) settings.Ypredef.value.filterNot(_ == "scala.Predef")
+      else settings.Ypredef.value
+
+    paths.map(path => rootMirror.getPackage(path))
+  }
```

Comment:
I think it's easier just to use Ypredef here. If you want to enforce exclusive use, you can move that into withPostSetHook.